Debug support for dapr k8s services and daprd (#1381)
dapr/docs: daprdocs/content/en/developing-applications/debugging/debug-k8s/debug-dapr-services.md

---
type: docs
title: "Debug Dapr services on Kubernetes"

> linkTitle: "Dapr services"

Reviewer: @msfussell @orizohar should we refer to these services as "Dapr services" or "Dapr control plane"?

> weight: 10000

Reviewer: ```suggestion
weight: 1000
```

description: "How to debug Dapr services on your Kubernetes cluster"
---

## Overview

> Sometimes it is necessary to understand what's going on in Dapr Kubernetes services, including `dapr-sidecar-injector`, `dapr-operator`, `dapr-placement`, and `dapr-sentry`, especially when you diagnose your Dapr application and wonder if there's something gets wrong in Dapr itself, or when you contribute to Dapr relevant to Kubernetes support.

Reviewer: ```suggestion
Sometimes it is necessary to understand what's going on in Dapr Kubernetes services, including `dapr-sidecar-injector`, `dapr-operator`, `dapr-placement`, and `dapr-sentry`, especially when you diagnose your Dapr application and wonder if there's something wrong in Dapr itself. Additionally, you may be developing a new feature for Dapr on Kubernetes and want to debug your code.
```

> If this is the case, Dapr has built-in Kubernetes debug support coming in to rescue.

Reviewer: ```suggestion
This guide will cover how to use Dapr debugging binaries to debug the Dapr services on your Kubernetes cluster.
```

> ## Debug Dapr Kubernetes services

Reviewer: ```suggestion
## Debugging Dapr Kubernetes services
```

Changing the tense to indicate this is a larger section with multiple steps

> Before read on, pls. refer [this guide]({{< ref kubernetes-deploy.md >}}) to learn how to deploy Dapr to your Kubernetes cluster.

Reviewer: ```suggestion
### Pre-requisites
- Familiarize yourself with [this guide]({{< ref kubernetes-deploy.md >}}) to learn how to deploy Dapr to your Kubernetes cluster.
- Setup your [dev environment](https://github.com/dapr/dapr/blob/master/docs/development/developing-dapr.md)
-  [Helm](https://github.com/helm/helm/releases)
```

### 1. Build Dapr debugging binaries

> In order to debug Dapr Kubernetes services, it's required to rebuild all Dapr binaries and Docker images to disable compiler optimization. To do this, pls. execute the following commands:

Reviewer: ```suggestion
In order to debug Dapr Kubernetes services, it's required to rebuild all Dapr binaries and Docker images to disable compiler optimization. To do this, execute the following commands:
```

```bash
git clone https://github.com/dapr/dapr.git
cd dapr
make release GOOS=linux GOARCH=amd64 DEBUG=1
```

> 

Reviewer: ```suggestion

>On Windows download [MingGW](https://sourceforge.net/projects/mingw/files/MinGW/Extension/make/mingw32-make-3.80-3/) and use `ming32-make.exe` instead of `make`.
```

In the above command, 'DEBUG' is specified  to '1' to disable compiler optimization. 'GOOS=linux' and 'GOARCH=amd64' are also necessary since the binaries will be packaged into Linux-based Docker image in the next step.

The binaries could be found under 'dist/linux_amd64/debug' sub-directory under the 'dapr' directory.

### 2. Build Dapr debugging Docker images

Use the following commands to package the debugging binaries into Docker images. Before this, you need to login your docker.io account, and if you don't have it yet, you may need to consider registering one from "https://hub.docker.com/".

```bash
export DAPR_TAG=dev
export DAPR_REGISTRY=<your docker.io id>
docker login
make docker-push DEBUG=1
```

Once the Dapr Docker images are built and pushed onto Docker hub, then you are ready to re-install Dapr in your Kubernetes cluster.

### 3. Install Dapr debugging binaries

If Dapr has already been installed in your Kubernetes cluster, uninstall it first:

```bash
dapr uninstall -k
```

> We will use 'helm' to install Dapr debugging binaries. For more information pls. check [Install with Helm]({{< ref "kubernetes-deploy.md#install-with-helm-advanced" >}}). In the following sections, we will use Dapr operator as an example to demonstrate how to configure, install, and debug Dapr services in a Kubernetes environment.

Reviewer: ```suggestion
We will use 'helm' to install Dapr debugging binaries. In the following sections, we will use Dapr operator as an example to demonstrate how to configure, install, and debug Dapr services in a Kubernetes environment.
```
